mathics/core/symbols.py

```diff
@@ -0,0 +1,783 @@
+    return (
+        isinstance(name, str)
+        and "`" in name
+        and not name.startswith("`")
+        and not name.endswith("`")
+        and "``" not in name
+    )
+
+
+def valid_context_name(ctx, allow_initial_backquote=False) -> bool:
+    return (
+        isinstance(ctx, str)
+        and ctx.endswith("`")
+        and "``" not in ctx
+        and (allow_initial_backquote or not ctx.startswith("`"))
+    )
+
+
+def ensure_context(name, context="System`") -> str:
+    assert isinstance(name, str)
+    assert name != ""
+    if "`" in name:
+        # Symbol has a context mark -> it came from the parser
+        assert fully_qualified_symbol_name(name)
+        return name
+    # Symbol came from Python code doing something like
+    # Expression('Plus', ...) -> use System` or more generally
+    # context + name
+    return context + name
+
+
+def strip_context(name) -> str:
+    if "`" in name:
+        return name[name.rindex("`") + 1 :]
+    return name
+
+
+class KeyComparable(object):
+    def get_sort_key(self):
+        raise NotImplementedError
+
+    def __lt__(self, other) -> bool:
+        return self.get_sort_key() < other.get_sort_key()
+
+    def __gt__(self, other) -> bool:
+        return self.get_sort_key() > other.get_sort_key()
+
+    def __le__(self, other) -> bool:
+        return self.get_sort_key() <= other.get_sort_key()
+
+    def __ge__(self, other) -> bool:
+        return self.get_sort_key() >= other.get_sort_key()
+
+    def __eq__(self, other) -> bool:
+        return (
+            hasattr(other, "get_sort_key")
+            and self.get_sort_key() == other.get_sort_key()
+        )
+
+    def __ne__(self, other) -> bool:
+        return (
+            not hasattr(other, "get_sort_key")
+        ) or self.get_sort_key() != other.get_sort_key()
+
+
+class BaseExpression(KeyComparable):
+    options: Any
+    pattern_sequence: bool
+    unformatted: Any
+    last_evaluated: Any
+
+    def __new__(cls, *args, **kwargs):
+        self = object.__new__(cls)
+        self.options = None
+        self.pattern_sequence = False
+        self.unformatted = self
+        self._cache = None
+        return self
+
+    def clear_cache(self):
+        self._cache = None
+
+    def equal2(self, rhs: Any) -> Optional[bool]:
+        """Mathics two-argument Equal (==)
+        returns True if self and rhs are identical.
+        """
+        if self.sameQ(rhs):
+            return True
+
+        # If the types are the same then we'll use the classes definition of == (or __eq__).
+        # Superclasses which need to specialized this behavior should redefine equal2()
+        #
+        # I would use `is` instead `==` here, to compare classes.
+        if type(self) is type(rhs):
+            return self == rhs
+        return None
+
+    def has_changed(self, definitions):
+        return True
+
+    def sequences(self):
+        return None
+
+    def flatten_sequence(self, evaluation) -> "BaseExpression":
```

Comment:
No quotes are needed for "BaseExpresion"

Reply:
flake8 complains if I remove the quotes...